testsuite/libffi.call/struct_by_value_big.c

```diff
@@ -5,6 +5,8 @@
    Originator:	From the original ffitest.c  */
 
 /* { dg-do run } */
+/* { dg-skip-if "structures larger than 16 bytes are passed by reference" { loongarch*-*-* } } */
```

Comment:
I have some doubt on this: IIRC on RISC-V the aggerates large than 2 * GRLEN are also passed by reference but the test is not disabled.  Should we fix the code actually handling the calling convention instead of papering over the issue?

Reply:
I don't think this test is valid considering [the current libffi policy about passing structs by value][1]:

> Note that argument values may be
modified by the callee (for instance, structs passed by value); the
burden of copying pass-by-value arguments is placed on the caller.

[1]:https://github.com/libffi/libffi/blob/master/doc/libffi.texi#L229

I guess this test also FAILs on RISC-V (but I don't have one RV board to test).

@atgreen should this test be removed completely or marked as XFAIL?

Reply:
> I don't think this test is valid considering [the current libffi policy about passing structs by value](https://github.com/libffi/libffi/blob/master/doc/libffi.texi#L229):
> 
> > Note that argument values may be
> > modified by the callee (for instance, structs passed by value); the
> > burden of copying pass-by-value arguments is placed on the caller.
> 
> I guess this test also FAILs on RISC-V (but I don't have one RV board to test).
> 
> @atgreen should this test be removed completely or marked as XFAIL?

According to #694, yes this also FAILs on RISC-V.

So which action should we take?  SKIP the test for targets like RISC-V and LoongArch, or just let it FAIL, or really fix this issue in the target code?  Need a clarification from a maintainer.

Reply:
I think libffi should make a copy and pass the pointer.  This means fixing aarch64 and risc-v as well.   I don't think we were thinking about this situation (large structs) when the documentation was written.   Please let loongarch64 show the way, and we'll address the other ports.

Reply:
> I think libffi should make a copy and pass the pointer. This means fixing aarch64 and risc-v as well. I don't think we were thinking about this situation (large structs) when the documentation was written. Please let loongarch64 show the way, and we'll address the other ports.

For AArch64 it seems enough to apply 482b37f00467325e3389bab322525099860dd9aa again.

Reply:
Yes, I agree. It looks like Andreas reverted it right after as he thought this was a bug in python.
Once we fix this in libffi, python will double-copy these large structs for a while, but I don't think we'll introduce any regressions or new errors.